[1.26] node: device-mgr: Fix recovery flow by ensuring healthy devices exist and pre-allocated devices are healthy

swatisehgal opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Cherry-pick of the following PR (/commits) on release-1.26:

1. https://github.com/kubernetes/kubernetes/pull/115926/commits/92e00203e027bf88ddc695fc17e80c913a6b178a of https://github.com/kubernetes/kubernetes/pull/115926: This commit contains sample device plugin refactoring.
2. https://github.com/kubernetes/kubernetes/pull/115107: This PR updates sample device plugin to enable the e2e node tests (or any other entity with full access to the node filesystem) to control the registration process. This feature is used in https://github.com/kubernetes/kubernetes/pull/114640 where e2e test is implemented to simulate scenarios where application pods requesting devices come up before the device plugin pod on node reboot/ kubelet restart.
3. https://github.com/kubernetes/kubernetes/pull/116240: This PR fixes a broken file path which got skipped as in the sample device plugin PR above.
4. https://github.com/kubernetes/kubernetes/pull/117057
5. https://github.com/kubernetes/kubernetes/pull/116376: This PR contains the core changes where we ensure that healthy devices exist and and pre-allocated devices are healthy. In addition to that, this PR contains e2e test to reproduce a scenario where application pod comes up before device plugin register itself to kubelet resulting in an admission error.  

**Rationale for the changes**
In case of node reboot/kubelet restart, the flow of events involves obtaining the state from the checkpoint file followed by setting the `healthDevices`/`unhealthyDevices` to its zero value. This is done to allow the device plugin to re-register itself so that capacity can be updated appropriately.

During the allocation phase, we need to check if the resources requested by the pod have been registered AND healthy devices are present on the node to be allocated.

Also we need to move this check above `needed==0` where needed is required - devices allocated to the container (which is obtained from the checkpoint file) because even in cases where no additional devices have to be allocated (as they were pre-allocated), we still need to make sure that the devices that were previously allocated are healthy.

For more details refer to the comment here: https://github.com/kubernetes/kubernetes/issues/109595#issuecomment-1359882423.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #109595 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
During device plugin allocation, resources requested by the pod can only be allocated if the device plugin has registered itself to kubelet AND healthy devices are present on the node to be allocated. If these conditions are not sattisfied, the pod would fail with `UnexpectedAdmissionError` error.
```

swatisehgal committed:
node: device-mgr: Handle recovery by checking if healthy devices exist

In case of node reboot/kubelet restart, the flow of events involves
obtaining the state from the checkpoint file followed by setting
the `healthDevices`/`unhealthyDevices` to its zero value. This is
done to allow the device plugin to re-register itself so that
capacity can be updated appropriately.

During the allocation phase, we need to check if the resources requested
by the pod have been registered AND healthy devices are present on
the node to be allocated.

Also we need to move this check above `needed==0` where needed is
required - devices allocated to the container (which is obtained from
the checkpoint file) because even in cases where no additional devices
have to be allocated (as they were pre-allocated), we still need to
make sure he devices that were previously allocated are healthy.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: sample device plugin: control registration process

Update the sample device plugin to enable the e2e node tests (or any
other entity with full access to the node filesystem) to control the
registration process. We add a new environment variable `REGISTER_CONTROL_FILE`.
The value of this variable must be a file which prevents the plugin
to register itself while it's present. Once removed, the plugin will
go on and complete the registration. The plugin will automatically
detect the parent directory on which the file resides and detect
deletions, unblocking the registration process. If the file is specified
but unaccessible, the plugin will fail. If the file is not specified,
the registration process will progress as usual and never pause.
The plugin will need read access to the parent directory.

This feature is useful because it is not possible to control the order
in which the pods are recovered after node reboot/kubelet restart.

In this approach, the testing environment will create a directory and
then a empty file to pause the registration process of the plugin.
Once pointed to that file, the plugin will start and wait for it to
be deleted. Only after the directory has been deleted,
the plugin would proceed to registration.

This feature is used in #114640 where e2e test is implemented to
simulate scenarios where application pods requesting devices come up before
the device plugin pod on node reboot/ kubelet restart.

Co-authored-by: Francesco Romani <fromani@redhat.com>
Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: sample device plugin: manifest to avoid registration

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: e2e: Implement End to end test

This commit reuses e2e tests implmented as part of https://github.com/kubernetes/kubernetes/pull/110729.
The commit is borrowed from the aforementioned PR as is to preserve
authorship. Subsequent commit will update the end to end test to
simulate the problem this PR is trying to solve by reproducing
the issue: 109595.

Co-authored-by: Francesco Romani <fromani@redhat.com>
Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: e2e: Update the e2e test to reproduce issue:109595

Breakdown of the steps implemented as part of this e2e test is as follows:
1. Create a file `registration` at path `/var/lib/kubelet/device-plugins/sample/`
2. Create sample device plugin with an environment variable with
   `REGISTER_CONTROL_FILE=/var/lib/kubelet/device-plugins/sample/registration` that
    waits for a client to delete the control file.
3. Trigger plugin registeration by deleting the abovementioned directory.
4. Create a test pod requesting devices exposed by the device plugin.
5. Stop kubelet.
6. Remove pods using CRI to ensure new pods are created after kubelet restart.
7. Restart kubelet.
8. Wait for the sample device plugin pod to be running. In this case,
   the registration is not triggered.
9. Ensure that resource capacity/allocatable exported by the device plugin is zero.
10. The test pod should fail with `UnexpectedAdmissionError`
11. Delete the test pod.
12. Delete the sample device plugin pod.
13. Remove `/var/lib/kubelet/device-plugins/sample/` and its content, the directory
    created to control registration

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

ffromani committed:
e2e: node: unify sample device plugin utilities

Start to consolidate the sample device plugin utility
and constants in a central place, because we need
to use it in different e2e tests.

Having a central dependency is better than a maze of
entangled e2e tests depending on each other helpers.

Signed-off-by: Francesco Romani <fromani@redhat.com>

swatisehgal committed:
node: device-mgr: e2e: adapt to sample device plugin refactoring

These updates are to adapt to the sample device plugin
refactoring done here: https://github.com/kubernetes/kubernetes/pull/115926/commits/92e00203e027bf88ddc695fc17e80c913a6b178a.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

bobbypage committed:
test: Fix path to e2e node sample device plugin

The existing path is incorrect (missing `sample-device-plugin`)
directory and thus causing test failures. The full path should be
`test/e2e/testing-manifests/sample-device-plugin/sample-device-plugin.yaml`.

Signed-off-by: David Porter <david@porter.me>

bobbypage commented:
/hold

This change may have resulted in test failures in https://github.com/kubernetes/kubernetes/issues/116340, we should triage those before attempting cherrypick.

swatisehgal committed:
node: device-plugin: e2e: Isolate test to pod restart scenario

Rather than testing out for both pod restart and kubelet restart,
we change the tests to just handle pod restart scenario.

Clarify the test purpose and add extra check to tighten the test.

We would be adding additional tests to cover kubelet restart scenarios
in subsequent commits.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>
Signed-off-by: Francesco Romani <fromani@redhat.com>

swatisehgal committed:
node: device-plugins: e2e: Refactor parse log to return string and error

Rather than only returning a string forcing us to log failure with
`framework.Fail`, we return a string and error to handle error cases
more conventionally. This enables us to use the `parseLog` function
inside `Eventually` and `Consistently` blocks, or in general to delegate
the error processing and enable better composability.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>
Co-authored-by: Francesco Romani <fromani@redhat.com>

swatisehgal committed:
node: device-plugins: e2e: s/devLen/expectedSampleDevsAmount

We rename to make the intent more explicit;
We make it global to be able to reuse the value all across the module
(e.g. to check the node readiness) later on.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>
Co-authored-by: Francesco Romani <fromani@redhat.com>

swatisehgal committed:
node: device-plugin: e2e: Update test description to make it explicit

Explicitly state that the test involves kubelet restart and device plugin
re-registration (no pod restart)

We remove the part of the code where we wait for the pod to restart as this
test case should no longer involve pod restart.

In addition to that, we use `waitForNodeReady` instead of `WaitForAllNodesSchedulable`
for ensuring that the node is ready for pods to be scheduled on it.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>
Co-authored-by: Francesco Romani <fromani@redhat.com>

swatisehgal committed:
node: device-plugin: e2e: Add test case for kubelet restart

Capture explicitly a test case pertaining to kubelet restart
but with no pod restart and device plugin re-registration.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-plugin: e2e: Annotate device check with error message

With this change the error message are more helpful and easier
to troubleshoot in case of test failures.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-plugin: e2e: Provide sleep intervals via constants

Based on whether the test case requires pod restart or not, the sleep
interval needs to be updated and we define constants to represent the two
sleep intervals that can be used in the corresponding test cases.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>
Co-authored-by: Francesco Romani <fromani@redhat.com>

swatisehgal committed:
node: device-plugin: e2e: Capture pod admission failure

This test captures that scenario where after kubelet restart,
application pod comes up and the device plugin pod hasn't re-registered
itself, the pod fails with admission error. It is worth noting that
once the device plugin pod has registered itself, another
application pod requesting devices ends up running
successfully.

For the test case where kubelet is restarted and device plugin
has re-registered without involving pod restart, since the
pod after kubelet restart ends up with admission error,
we cannot be certain the device that the second pod (pod2) would
get. As long as, it gets a device we consider the test to pass.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-plugin: add node reboot test scenario

Add a test suit to simulate node reboot (achieved by removing pods
using CRI API before kubelet is restarted).

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-plugin: e2e: Additional test cases

Additional test cases added:
Keeps device plugin assignments across pod and kubelet restarts (no device plugin re-registration)
Keeps device plugin assignments after the device plugin has re-registered (no kubelet or pod restart)

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/restest

swatisehgal commented:
/retest

swatisehgal commented:
/test pull-kubernetes-node-e2e-containerd
/test pull-kubernetes-e2e-gce-serial

swatisehgal commented:
/hold

cici37 commented:
/remove-sig api-machinery

swatisehgal commented:
/remove-sig architecture auth cli cloud-provider cluster-lifecycle instrumentation storage

ffromani commented:
/triage accepted
/priority important-longterm

ffromani commented:
/kind bug

swatisehgal commented:
/test pull-kubernetes-e2e-gce-serial

swatisehgal commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd-serial

ffromani commented:
/cc

I can review once the WIP is lifted

rphillips commented:
/lgtm

ffromani commented:
/lgtm

disclosure: I co-authored a large part of the changes.
The backport seems good. While this is a large bakport, 8/18 commits (~45%) are trivial cleanups or refactoring. The core changes are small and self contained, and a very significant effort here is spent backporting test to ensure the correctness of the changes and the surrounding code. The 1.26 codebase is still quite similar to 1.27/main branch.

SergeyKanzhelev commented:
/test

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

swatisehgal commented:
> @swatisehgal changes looks good. As @ffromani said a lot of refactoring, but generally ok.
> 
> Looking at bugs we are fixing - those are marked as priority/important-longterm. Typically not something that needs cherry-pick. But I understand the priority and the need for cherry pick was discussed? Can you please link the dicussion.

The issue was discussed in the SIG Node meetings on [26/01/2023](https://docs.google.com/document/d/1Ne57gvidMEWXR70OxxnRkYquAoMpt56o75oZtg-OeBg/edit#bookmark=id.lda63gwg1uyy) but we didn't discuss if the bugfix would be a cherry-pick candidate for the previous maintenance releases. Since device manager is a GA feature, I assumed that we have a strong case to go ahad with the backport process which I am realizing is not ideal. Its best we discuss it with others in the community and get an explicit ACK on this. I have added an agenda item for this to be discussed in the upcoming SIG Node meeting. 
> Also, which of test jobs executing this code?

Both [pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial ](https://testgrid.k8s.io/sig-node-containerd#cos-cgroupv1-containerd-node-e2e-serial) and [pull-kubernetes-cos-cgroupv1-containerd-node-e2e](https://testgrid.k8s.io/sig-node-containerd#cos-cgroupv2-containerd-node-e2e-serial) should run these tests but they don't seems to be available here. Would you know how I can enable them here?

dchen1107 commented:
Reviewed the diff between this one and the backport for 1.27 based on the guidance offered by @ffromani: https://github.com/kubernetes/kubernetes/pull/116337#issuecomment-1535893004

/lgtm
/approve from SIG Node perspective

xmudrii reviewed (approved):
/lgtm
/approve
